11x-new-kafka-roller.md

```diff
@@ -0,0 +1,263 @@
+
+`NOT_READY` nodes will be restarted if they have a restart reason and have not been restarted yet.
+If it is not ready after being restarted already, we don't want to restart any other nodes to avoid taking down more nodes.
+
+`READY` nodes will be restarted if they have a restart reason.
+If they don't have a restart reason but need to be reconfigured, they will be reconfigured.
+If no reconfiguration is needed, then no action will be taken on these nodes.
+
+If all nodes reach the desired state, the reconciliation will succeed, otherwise it will fail.
+There are also some conditions that could result in early termination of the process and immediately failing the reconciliation.
+This will be explained more later.
+
+### High level flow diagram describing the flow of the states
+![The new roller flow](./images/06x-new-roller-flow.png)
+
+### State machine cycles
+
+As previously mentioned, the process for nodes will be repeated unless the maximum attempt is reached, in which case the reconciliation fails.
+The maximum number of attempts is hard-coded to 10 in the current roller.
+It will be the same for the new roller, however this value will be further broken down.
+The new roller will add 2 other hard-coded maximum attempt values.
+One is for the maximum number of restarts and one is for the maximum number of reconfigurations that can be taken on each node.
+This is because we want to limit how many times a node can be restarted in each reconciliation because restarting a node 10 times is not productive.
+Also, in the current roller, if we failed to reconfigure a node, we immediately restart it.
+Reconfiguration can fail sometimes due to transitive error so it would be useful to retry the reconfiguration a few times before we decide to restart a node.
+
+In each reconciliation, the number of attempts is tracked for each node.
+The number of attempts is how many times the overall process is repeated per node because of not reaching the desired state and the number of restarts is how many times a node is actually restarted.
+If any node has reached the maximum number of attempts or restarts, the reconciliation will fail.
+If the maximum number of reconfiguration is reached, then the node will be marked to restart but will not fail the reconciliation.
+When a new reconciliation starts, the tracked number of actions taken on nodes will be reset.
+
+The current roller also fails the reconciliation in the following situation to avoid continuing the process on more nodes:
+- Cannot connect to KafkaAgent to check broker state metrics (KafkaAgent is crucial when determining safety before rolling nodes, so there is no point to try to progress further, if we cannot connect to the KafkaAgent).
+- Pod is not Running and does not have an old revision (this will prevent the roller from restarting more nodes, which could result in making them stuck as well and bring down the entire cluster).
```

Comment:
It's not clear whether this behaviour is being kept in the new KafkaRoller or not